In [ ]:
from scraper_actualizado import Scraper

URLS = [
    {
        "url": "https://es.wikisource.org/wiki/Historia_de_un_amor_turbio",
        "multipagina": False
    },
    {
        "url": "https://es.wikisource.org/wiki/A_flor_de_piel",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/A_fuego_lento",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/A_los_pies_de_Venus",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/A_prueba",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/A_todo_honor",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Adem%C3%A1s_del_frac",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Aita_Tettauen",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Al_amor_de_los_tizones",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Al_primer_vuelo",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Amadeo_I",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Amalia",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Amistad_funesta",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Ana_Karenina",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/%C3%81ngel_Guerra_(Gald%C3%B3s)",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Arroz_y_tartana",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/As%C3%AD_paga_el_diablo",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Aura_o_las_violetas",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Aurora_roja",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Baile_y_cochino",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Bail%C3%A9n_(P%C3%A9rez_Gald%C3%B3s)",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Blasones_y_talegas",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Bodas_reales",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Brenda",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Buenos_Aires_en_el_a%C3%B1o_2080",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Bug-Jargal",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/C%C3%A1diz",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Camila_o_la_virtud_triunfante",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/C%C3%A1novas",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Caramur%C3%BA",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Carlos_VI_en_la_R%C3%A1pita",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Ca%C3%B1as_y_barro",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Cel%C3%ADn",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Cinco_semanas_en_globo",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Clemencia_(Altamirano)",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Clemencia_(Caballero)",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Crimen_y_castigo_(tr._an%C3%B3nima)",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Cr%C3%B3nica_del_reinado_de_Carlos_IX",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Cuesti%C3%B3n_de_ambiente",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Cumand%C3%A1",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/De_Cartago_a_Sagunto",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/De_O%C3%B1ate_a_la_Granja",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/De_tal_palo,_tal_astilla",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Del_fr%C3%ADo_al_fuego",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Don_Gonzalo_Gonz%C3%A1lez_de_la_Gonzalera",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Don_Segundo_Sombra",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Do%C3%B1a_Luz",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Do%C3%B1a_Milagros",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Do%C3%B1a_Perfecta",
        "multipagina": True
    },
    {
        "url": "https://es.wikisource.org/wiki/Dulce_due%C3%B1o_(Pardo_Baz%C3%A1n)",
        "multipagina": True
    }
]

NOMBRE_DE_CARPETA = "datos"

CABECERAS = {
    "User-Agent": "LabPLN-2026/1.0 (contacto: tu-mail@ejemplo.com)"
}

STOPWORDS = r"c:\Ciencia De Datos\NLP\pair_programming_ia\stopwords_es.txt"


def cargar_stopwords(ruta_archivo):
    """Carga el archivo de stopwords y devuelve un conjunto de palabras."""
    with open(ruta_archivo, "r", encoding="utf-8") as f:
        return {
            linea.strip().lower()
            for linea in f
            if linea.strip()
        }


def inicializar_scraper(url, nombre_archivo):
    """
    Inicializa el scraper, verifica robots.txt,
    obtiene la página principal y guarda su HTML.
    """
    scraper = Scraper(
        url,
        headers=CABECERAS,
        nombre_archivo=nombre_archivo,
        carpeta_destino=NOMBRE_DE_CARPETA
    )

    permitido = scraper.permite_scrapeo()

    if permitido:
        print(f"Scrapeo permitido por robots.txt: {permitido}")
        scraper.actualizo_csv(url, "robots", "OK")

        respuesta = scraper.obtener_pagina()

        if respuesta:
            print("Scrapeo exitoso")
            scraper.guardar_pagina()
            scraper.actualizo_csv(url, "extraccion", "OK")
            return scraper

        print("No se pudo obtener la página")
        scraper.actualizo_csv(url, "extraccion", "Error")
        return False

    print("Scrapeo no permitido")
    scraper.actualizo_csv(url, "robots", "No permitido")

    print(
        "Forzar scrapeo sin autorización de robots.txt "
        "(no recomendado) s/n:"
    )
    respuesta = input().lower()

    if respuesta == "s":
        print("Forzando scrapeo sin autorización de robots.txt")

        respuesta = scraper.obtener_pagina()

        if respuesta:
            print("Scrapeo exitoso")
            scraper.guardar_pagina()
            scraper.actualizo_csv(url, "extraccion", "OK")
            return scraper

    print("Scrapeo cancelado")
    scraper.actualizo_csv(url, "extraccion", "Cancelado")

    return False


def contar_palabra(scraper, palabra):
    """Cuenta cuántas veces aparece una palabra en la página principal."""
    cantidad = scraper.contar_palabra(palabra)

    print(
        f"La palabra '{palabra}' aparece "
        f"{cantidad} veces en la página."
    )


def main():
    """Recorre todas las obras definidas en URLS y arma el corpus."""

    stopwords = cargar_stopwords(STOPWORDS)

    for i, item in enumerate(URLS, start=1):

        url = item["url"]
        multipagina = item["multipagina"]

        print("\n" + "=" * 60)
        print(f"Procesando URL {i}:")
        print(url)
        print(f"Multipágina: {multipagina}")
        print("=" * 60)

        nombre_archivo = f"pagina_{i}.html"

        scraper = inicializar_scraper(
            url,
            nombre_archivo
        )

        if not scraper:
            print(
                "Esta URL no pudo ser procesada. "
                "Se continúa con la siguiente."
            )
            continue

        scraper.actualizo_csv(
            url,
            "multipagina",
            "Sí" if multipagina else "No"
        )

        # Los metadatos se extraen siempre de la página principal.
        datos_corpus = scraper.obtener_datos_corpus()
        
       
        scraper.actualizo_csv(
            url,
            "ruta",
            str(scraper.ruta_archivo)
        )

        scraper.actualizo_csv(
            url,
            "Titulo de la página",
            datos_corpus.get("title", "")
        )

        scraper.actualizo_csv(
            url,
            "Autoría de la página",
            datos_corpus.get("author", "")
        )

        scraper.actualizo_csv(
            url,
            "Fecha de publicación de la página",
            datos_corpus.get("date", "")
        )
        
        fecha_publicacion_obra = scraper.obtener_fecha_publicacion_obra()
        
        scraper.actualizo_csv(
            url,
            "Año de publicación de la Obra",
            fecha_publicacion_obra
        )
        
        print(f"Fecha de publicación obtenida: {fecha_publicacion_obra}")
        
        scraper.actualizo_csv(
            url,
            "Medio",
            datos_corpus.get("hostname", "")
        )

        # --------------------------------------------------------
        # ACÁ SE DECIDE EXPLÍCITAMENTE QUÉ TIPO DE SCRAPING USAR.
        # --------------------------------------------------------
        if multipagina:
            texto_corpus = scraper.obtener_texto_multipagina()
            
        else:
            texto_corpus = datos_corpus.get("text", "")

            
        if not texto_corpus:
            print("No se pudo extraer texto del corpus.")
            scraper.actualizo_csv(
                url,
                "extraccion",
                "Sin texto"
            )
            continue

        datos_normalizados = scraper.normalizar(
            texto_corpus
        )

        scraper.actualizo_csv(
            url,
            "normalizacion",
            "OK"
        )
        resultado = scraper.tokenizar_documentos([datos_normalizados])
        print("Resultado de la tokenización:", resultado)
        scraper.actualizo_csv(
            url,
            "Cant-Tokens",
            resultado
        )
        
        scraper.actualizo_csv(
            url,
            "Licencia",
            datos_corpus.get("license", "")
        )
        
        datos_sin_stopwords = scraper.palabras_significativas(
            texto_corpus,
            stopwords
        )

        print(
            "Datos del corpus sin stopwords: "
            f"{datos_sin_stopwords[:100]}..."
        )

        scraper.guardar_datos_corpus(
            datos_normalizados,
            tipo_dato="texto",
            tipo_archivo="normalizado"
        )
        
        scraper.actualizo_csv(
            url,
            "guardado",
            "OK"
        )    

if __name__ == "__main__":
    main()


In [ ]:
from scraper_actualizado import Scraper as Scraper_1

scraper_1 = Scraper_1(
    url="https://es.wikisource.org",
    carpeta_destino="datos"
)
scraper_1.generar_corpus_jsonl()      

In [5]:
import json

def analizar_dispersion_longitud(ruta_corpus="datos/corpus.jsonl"):
    """
    Calcula la cantidad de tokens de cada documento
    utilizando len(texto.split()).
    """

    cantidades = []

    with open(ruta_corpus, "r", encoding="utf-8") as archivo:

        for linea in archivo:

            documento = json.loads(linea)

            cantidad_tokens = len(documento["texto"].split())

            cantidades.append(cantidad_tokens)

            print(
                f"{documento['id']} | "
                f"{documento['titulo']} | "
                f"{cantidad_tokens} tokens"
            )

    print("\n--- RESUMEN ---")
    print(f"Cantidad de documentos: {len(cantidades)}")
    print(f"Mínimo: {min(cantidades)} tokens")
    print(f"Máximo: {max(cantidades)} tokens")
    print(f"Diferencia: {max(cantidades) - min(cantidades)} tokens")
    print(f"Promedio: {sum(cantidades) / len(cantidades):.2f} tokens")
analizar_dispersion_longitud()    

doc_01 | Caramurú - Wikisource | 48766 tokens
doc_02 | Doña Luz - Wikisource | 59381 tokens
doc_03 | A flor de piel - Wikisource | 60849 tokens
doc_04 | Don Segundo Sombra - Wikisource | 61274 tokens
doc_05 | A fuego lento - Wikisource | 64006 tokens
doc_06 | Dulce dueño (Pardo Bazán) - Wikisource | 66584 tokens
doc_07 | Crónica del reinado de Carlos IX - Wikisource | 66868 tokens
doc_08 | Doña Milagros - Wikisource | 67790 tokens
doc_09 | De Cartago a Sagunto - Wikisource | 68103 tokens
doc_10 | Aurora roja - Wikisource | 69763 tokens
doc_11 | Carlos VI en la Rápita - Wikisource | 70053 tokens
doc_12 | Cañas y barro - Wikisource | 74953 tokens

--- RESUMEN ---
Cantidad de documentos: 12
Mínimo: 48766 tokens
Máximo: 74953 tokens
Diferencia: 26187 tokens
Promedio: 64865.83 tokens


In [4]:
import json
import re

def revisar_artefactos(ruta_corpus="datos/corpus.jsonl"):
    
    with open(ruta_corpus, "r", encoding="utf-8") as archivo:

        for linea in archivo:
            documento = json.loads(linea)

            texto = documento["texto"]

            html = re.findall(r"<[^>]+>", texto)
            javascript = re.findall(
                r"\b(function|var|let|const|script)\b",
                texto,
                flags=re.IGNORECASE
            )

            espacios_duros = texto.count("\xa0")

            print(
                f"{documento['id']} | "
                f"{documento['titulo']}"
            )

            print(f"Espacios duros: {espacios_duros}")
            print(f"Restos HTML: {len(html)}")
            print(f"Posibles restos JS: {len(javascript)}")
            print("-" * 50)
revisar_artefactos()            

doc_01 | Caramurú - Wikisource
Espacios duros: 0
Restos HTML: 0
Posibles restos JS: 0
--------------------------------------------------
doc_02 | Doña Luz - Wikisource
Espacios duros: 0
Restos HTML: 0
Posibles restos JS: 0
--------------------------------------------------
doc_03 | A flor de piel - Wikisource
Espacios duros: 0
Restos HTML: 0
Posibles restos JS: 0
--------------------------------------------------
doc_04 | Don Segundo Sombra - Wikisource
Espacios duros: 0
Restos HTML: 0
Posibles restos JS: 0
--------------------------------------------------
doc_05 | A fuego lento - Wikisource
Espacios duros: 0
Restos HTML: 0
Posibles restos JS: 0
--------------------------------------------------
doc_06 | Dulce dueño (Pardo Bazán) - Wikisource
Espacios duros: 0
Restos HTML: 0
Posibles restos JS: 0
--------------------------------------------------
doc_07 | Crónica del reinado de Carlos IX - Wikisource
Espacios duros: 0
Restos HTML: 0
Posibles restos JS: 1
------------------------------

In [6]:
import json
import spacy
import pandas as pd

# Cargar modelo de español
nlp = spacy.load("es_core_news_sm")

# Ruta del corpus
ruta_corpus = "datos/corpus.jsonl"

resultados = []

with open(ruta_corpus, "r", encoding="utf-8") as archivo:

    for linea in archivo:

        documento = json.loads(linea)

        texto = documento["texto"]

        # Procesamiento lingüístico con spaCy
        doc = nlp(texto)

        # Solo tokens que sean palabras:
        # excluimos puntuación y espacios
        palabras = [
            token
            for token in doc
            if not token.is_punct and not token.is_space
        ]

        total_palabras = len(palabras)

        # Conteo POS
        cantidad_adjetivos = sum(
            1 for token in palabras
            if token.pos_ == "ADJ"
        )

        cantidad_adverbios = sum(
            1 for token in palabras
            if token.pos_ == "ADV"
        )

        # Frecuencias relativas
        frecuencia_adj = (
            cantidad_adjetivos / total_palabras
            if total_palabras > 0 else 0
        )

        frecuencia_adv = (
            cantidad_adverbios / total_palabras
            if total_palabras > 0 else 0
        )

        resultados.append({
            "id": documento["id"],
            "titulo": documento["titulo"],
            "total_palabras": total_palabras,
            "adjetivos": cantidad_adjetivos,
            "adverbios": cantidad_adverbios,
            "fr_adj": frecuencia_adj,
            "fr_adv": frecuencia_adv,
            "adj_mayor_adv": frecuencia_adj > frecuencia_adv
        })


# Convertir resultados a DataFrame
df_resultados = pd.DataFrame(resultados)

# Mostrar porcentajes con más claridad
df_resultados["fr_adj_pct"] = (
    df_resultados["fr_adj"] * 100
).round(2)

df_resultados["fr_adv_pct"] = (
    df_resultados["fr_adv"] * 100
).round(2)

# Columnas finales para visualizar
columnas_mostrar = [
    "id",
    "titulo",
    "total_palabras",
    "adjetivos",
    "adverbios",
    "fr_adj_pct",
    "fr_adv_pct",
    "adj_mayor_adv"
]

display(df_resultados[columnas_mostrar])

,id,titulo,total_palabras,adjetivos,adverbios,fr_adj_pct,fr_adv_pct,adj_mayor_adv
0,doc_01,Caramurú - Wikisource,48782,3802,2047,7.79,4.20,True
1,doc_02,Doña Luz - Wikisource,59401,4073,3610,6.86,6.08,True
2,doc_03,A flor de piel - Wikisource,60802,5886,2669,9.68,4.39,True
3,doc_04,Don Segundo Sombra - Wikisource,61205,3984,2859,6.51,4.67,True
4,doc_05,A fuego lento - Wikisource,64006,4846,2884,7.57,4.51,True
5,doc_06,Dulce dueño (Pardo Bazán) - Wikisource,66459,5521,3346,8.31,5.03,True
6,doc_07,Crónica del reinado de Carlos IX - Wikisource,65207,4213,3436,6.46,5.27,True
7,doc_08,Doña Milagros - Wikisource,67788,4743,3731,7.00,5.50,True
8,doc_09,De Cartago a Sagunto - Wikisource,68132,5022,2835,7.37,4.16,True
9,doc_10,Aurora roja - Wikisource,69794,4590,3381,6.58,4.84,True


In [7]:
documentos_cumplen = df_resultados[
    "adj_mayor_adv"
].sum()

total_documentos = len(df_resultados)

print(
    f"Documentos donde ADJ > ADV: "
    f"{documentos_cumplen} de {total_documentos}"
)

print(
    f"Documentos donde ADJ <= ADV: "
    f"{total_documentos - documentos_cumplen}"
)

promedio_adj = (
    df_resultados["fr_adj"].mean() * 100
)

promedio_adv = (
    df_resultados["fr_adv"].mean() * 100
)

print(
    f"Frecuencia relativa media de adjetivos: "
    f"{promedio_adj:.2f}%"
)

print(
    f"Frecuencia relativa media de adverbios: "
    f"{promedio_adv:.2f}%"
)

Documentos donde ADJ > ADV: 12 de 12
Documentos donde ADJ <= ADV: 0
Frecuencia relativa media de adjetivos: 7.37%
Frecuencia relativa media de adverbios: 4.83%


Fernando - Puntos 3.4, 3.5 y 3.6


In [4]:
STOPWORDS_mio = r"c:\Ciencia De Datos\NLP\pair_programming_ia\stopwords_es.txt"


def cargar_stopwords(ruta_archivo):
    """Carga el archivo de stopwords y devuelve un conjunto de palabras."""
    with open(ruta_archivo, "r", encoding="utf-8") as f:
        return {
            linea.strip().lower()
            for linea in f
            if linea.strip()
        }

stopwords_m = cargar_stopwords(STOPWORDS_mio)

In [10]:
import json
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer

nro_de_texto = 3
nro_de_palabra = 2

# 1. Cargamos el corpus desde el archivo corpus.jsonl
mio_texts = []
mio_metadata = []
with open('corpus.jsonl', 'r', encoding='utf-8') as f:
    for mio_line in f:
        mio_item = json.loads(mio_line)
        mio_texts.append(mio_item['texto'])
        mio_metadata.append(mio_item)

mio_df_meta = pd.DataFrame(mio_metadata)
mio_df_meta['token_count'] = [len(t.split()) for t in mio_texts]

# 2. Vectorizamos utilizando la variable 'stopword_es' del código previo
#mio_cv = CountVectorizer(stop_words=stopwords_m, token_pattern=r'\b[a-záéíóúñ]{4,}\b')
mio_cv = CountVectorizer(stop_words=list(stopwords_m), token_pattern=r'\b[a-záéíóúñ]{4,}\b')
mio_X_counts = mio_cv.fit_transform(mio_texts)
mio_feature_names = mio_cv.get_feature_names_out()

# 3. Elegimos un texto del corpus
mio_idx_texto_elegido = nro_de_texto
mio_titulo_texto = mio_df_meta.loc[mio_idx_texto_elegido, 'titulo']
print(f'Texto seleccionado para la auditoría: {mio_titulo_texto}')

# Obtenemos las frecuencias absolutas para ese texto específico
mio_doc_vector = mio_X_counts[mio_idx_texto_elegido].toarray().flatten()
mio_df_top_doc = pd.DataFrame(
    {'palabra': mio_feature_names, 'conteo_absoluto': mio_doc_vector}
).sort_values(by='conteo_absoluto', ascending=False)

print('\nTop 5 palabras más frecuentes en este texto:')
display(mio_df_top_doc.head(5))

# 4. Seleccionamos una palabra representativa de ese top
mio_palabra_auditoria = mio_df_top_doc.iloc[nro_de_palabra]['palabra'] 
# Si querés cambiarla manual, podés poner por ejemplo: mio_df_top_doc.iloc[1]['palabra']
mio_idx_palabra = mio_cv.vocabulary_[mio_palabra_auditoria]

print(f"\nPalabra elegida para auditar en el corpus: '{mio_palabra_auditoria}'")

# 5. Armamos la tabla comparativa de todo el corpus
mio_df_audit_34 = pd.DataFrame({
    'Título': mio_df_meta['titulo'],
    'Tokens': mio_df_meta['token_count'],
    'Conteo_Crudo': mio_X_counts[:, mio_idx_palabra].toarray().flatten(),
})

# Calculamos la frecuencia relativa (ej. cada 10.000 palabras)
mio_df_audit_34['Frecuencia_Relativa_10k'] = (
    mio_df_audit_34['Conteo_Crudo'] / mio_df_audit_34['Tokens']
) * 10000

print('\n--- Resultados de la Auditoría (Ordenado por mayor conteo crudo) ---')
display(mio_df_audit_34.sort_values(by='Conteo_Crudo', ascending=False))



Texto seleccionado para la auditoría: Don Segundo Sombra - Wikisource

Top 5 palabras más frecuentes en este texto:


,palabra,conteo_absoluto
26804,había,287
46672,segundo,182
27545,hombre,141
17996,dijo,123
7505,bien,116



Palabra elegida para auditar en el corpus: 'hombre'

--- Resultados de la Auditoría (Ordenado por mayor conteo crudo) ---


,Título,Tokens,Conteo_Crudo,Frecuencia_Relativa_10k
9,Aurora roja - Wikisource,69763,186,26.661697
3,Don Segundo Sombra - Wikisource,61274,141,23.011391
6,Crónica del reinado de Carlos IX - Wikisource,66868,118,17.646707
4,A fuego lento - Wikisource,64006,105,16.404712
11,Cañas y barro - Wikisource,74953,81,10.806772
10,Carlos VI en la Rápita - Wikisource,70053,69,9.849685
7,Doña Milagros - Wikisource,67790,68,10.030978
0,Caramurú - Wikisource,48766,65,13.328959
8,De Cartago a Sagunto - Wikisource,68103,54,7.929166
1,Doña Luz - Wikisource,59381,53,8.925414


Conclusión sobre la auditoría de frecuencia (Punto 3.4):

Al analizar la distribución de la palabra «hombre» en el corpus, inicialmente observamos que los textos con mayor conteo absoluto tienden a coincidir en líneas generales con los de mayor longitud en tokens. Sin embargo, al calcular la frecuencia relativa normalizada cada 10.000 palabras, se verifica que las diferencias en las menciones no responden exclusivamente a una trampa de extensión o "falacia del texto largo" (dado que nuestros textos poseen longitudes relativamente homogéneas entre 48k y 74k tokens).

Un caso claro se observa al comparar Aurora roja y Don Segundo Sombra: aunque este último tiene una cantidad menor de tokens (61.274 frente a 69.763), su frecuencia relativa de la palabra «hombre» es alta (23.01), ubicándose muy cerca de la obra que encabeza la lista. Esto demuestra que la normalización cumple su función metodológica al aislar el peso estilístico y temático real de cada obra, permitiendo ver que la densidad de uso del término varía por decisiones discursivas del autor o género, y no por el mero volumen de texto.

REVISAR: como entiendo que el objetivo de este punto es demostrar cómo funciona la frecuencia relativa frente al conteo absoluto y ver el impacto de la longitud del texto, no usé una palabra que se adjetivo o adverbio, y se podría elegir cualquier palabra representativa del corpus (como probamos con «gaucho» o «hombre»), dado el objetivo

Punto 3.5
Auditoría TTR
Calcular TTR por texto y cruzarlo con longitud para verificar la falacia de que los textos cortos parecen más ricos.
resultado en Tabla/gráfico o análisis TTR vs longitud.

In [2]:
import json
import pandas as pd

# 1. Cargamos directamente el corpus para asegurarnos de tener los textos a mano
datos_corpus = []
with open('datos/corpus.jsonl', 'r', encoding='utf-8') as f:
    for linea in f:
        datos_corpus.append(json.loads(linea))

df_corpus_ttr = pd.DataFrame(datos_corpus)

# 2. Listas para la auditoría de TTR
resultados_ttr = []

# 3. Recorremos cada documento
for i, row in df_corpus_ttr.iterrows():
    texto = row['texto']
    titulo = row['titulo']
    
    # Tokenización estándar por espacios (sin filtrar stopwords para el TTR clásico)
    tokens_ttr = texto.split()
    total_tokens_ttr = len(tokens_ttr)
    
    # Types únicos (pasados a minúsculas)
    types_unicos = set(token.lower() for token in tokens_ttr)
    total_types = len(types_unicos)
    
    # Cálculo del Type-Token Ratio
    ttr_valor = total_types / total_tokens_ttr if total_tokens_ttr > 0 else 0
    
    resultados_ttr.append({
        'Título': titulo,
        'Tokens': total_tokens_ttr,
        'Types_Unicos': total_types,
        'TTR': ttr_valor
    })

# DataFrame exclusivo para el análisis de TTR ordenado por cantidad de tokens
df_auditoria_ttr = pd.DataFrame(resultados_ttr)

print("--- Auditoría TTR vs Longitud ---")
display(df_auditoria_ttr.sort_values(by='Tokens', ascending=True))

--- Auditoría TTR vs Longitud ---


,Título,Tokens,Types_Unicos,TTR
0,Caramurú - Wikisource,48766,12643,0.259258
1,Doña Luz - Wikisource,59381,12722,0.214244
2,A flor de piel - Wikisource,60849,16587,0.272593
3,Don Segundo Sombra - Wikisource,61274,15012,0.244998
4,A fuego lento - Wikisource,64006,17035,0.266147
5,Dulce dueño (Pardo Bazán) - Wikisource,66584,19834,0.297879
6,Crónica del reinado de Carlos IX - Wikisource,66868,14053,0.210160
7,Doña Milagros - Wikisource,67790,18232,0.268948
8,De Cartago a Sagunto - Wikisource,68103,17312,0.254203
9,Aurora roja - Wikisource,69763,15159,0.217293


Observaciones:
Tokens: Son todas las palabras que componen el texto, una por una, contando absolutamente todas las apariciones (incluso si se repiten). Por ejemplo, si la palabra "el" aparece 50 veces en una novela, cuenta como 50 tokens distintos dentro del total de palabras del libro.

Types_Unicos (o Types): Son las palabras diferentes o únicas que forman el vocabulario de ese texto, ignorando por completo cuántas veces se repiten. Siguiendo el ejemplo anterior, no importa si "el" apareció 50 veces; como type solo cuenta como 1 palabra única en el inventario del vocabulario.

TTR (Type-Token Ratio): Es la división matemática entre ambos (Types / Tokens). Esta métrica mide la variedad o "riqueza léxica" del texto.
Entonces:
    Si un texto tiene un TTR cercano a 1 (o 100%), significa que casi todas sus palabras son diferentes y casi no hay repeticiones (típico de textos muy cortos).
    Si tiene un TTR bajo (cercano a 0), significa que el texto recicla mucho las mismas palabras (artículos, verbos, pronombres) a lo largo de miles de tokens, lo cual es normal en novelas más extensas.

Conclusión:
Al analizar los resultados de la auditoría de TTR en relación con la longitud de los textos, se verifica que los textos ordenados por cantidad de tokens, de menor a mayor, muestran una relación inversamente proporcional con su TTR.
Entonces:
Las obras con menor cantidad de tokens presentan los valores de TTR más elevados (acercándose a proporciones de vocabulario único más altas), mientras que los textos más extensos registran un TTR considerablemente menor. Esto demuestra empíricamente que el TTR sin normalizar penaliza a los textos largos —debido a la constante repetición natural de palabras funcionales y estructurales como artículos, preposiciones y verbos auxiliares—, generando el sesgo erróneo de que los textos breves poseen una mayor riqueza o diversidad léxica por el simple hecho de ser cortos.

Punto 3.6:
Auditoría TF-IDF
Explicar y comprobar en el corpus el efecto de smooth_idf=True y la norma L2; revisar por qué términos presentes en todos los documentos no necesariamente dan 0.
Comprobación sklearn vs cálculo/explicación de clase.

In [4]:
import json
from sklearn.feature_extraction.text import TfidfVectorizer
import pandas as pd
import numpy as np

# 1. Cargamos directamente el corpus del archivo para asegurar la independencia de la celda
datos_corpus = []
with open('datos/corpus.jsonl', 'r', encoding='utf-8') as f:
    for linea in f:
        datos_corpus.append(json.loads(linea))

df_corpus = pd.DataFrame(datos_corpus)

textos = df_corpus['texto'].tolist()
titulos = df_corpus['titulo'].tolist()

# 2. Vectorización TF-IDF con los parámetros exigidos por la auditoría (smooth_idf=True y norm='l2')
vectorizer = TfidfVectorizer(smooth_idf=True, norm='l2')
matriz_tfidf = vectorizer.fit_transform(textos)

# 3. Comprobación de smooth_idf: analizamos los valores IDF calculados por sklearn
df_idf = pd.DataFrame({
    'Palabra': vectorizer.get_feature_names_out(),
    'IDF': vectorizer.idf_
})

print("--- Auditoría TF-IDF: Comprobación de smooth_idf (ningún IDF da 0) ---")
display(df_idf.sort_values(by='IDF', ascending=True).head(10))

# 4. Comprobación de la Norma L2 por documento (cada vector de fila debe normalizarse a 1.0)
normas_filas = np.linalg.norm(matriz_tfidf.toarray(), axis=1)
df_normas = pd.DataFrame({'Título': titulos, 'Norma_L2': normas_filas})

print("\n--- Auditoría TF-IDF: Verificación de Normalización L2 (vectores unitarios) ---")
display(df_normas)

--- Auditoría TF-IDF: Comprobación de smooth_idf (ningún IDF da 0) ---


,Palabra,IDF
24260,falta,1.0
54505,última,1.0
40665,porque,1.0
40634,por,1.0
40754,posesión,1.0
37867,palabra,1.0
13266,contar,1.0
8793,cadáver,1.0
53221,vi,1.0
53214,veía,1.0



--- Auditoría TF-IDF: Verificación de Normalización L2 (vectores unitarios) ---


,Título,Norma_L2
0,Caramurú - Wikisource,1.0
1,Doña Luz - Wikisource,1.0
2,A flor de piel - Wikisource,1.0
3,Don Segundo Sombra - Wikisource,1.0
4,A fuego lento - Wikisource,1.0
5,Dulce dueño (Pardo Bazán) - Wikisource,1.0
6,Crónica del reinado de Carlos IX - Wikisource,1.0
7,Doña Milagros - Wikisource,1.0
8,De Cartago a Sagunto - Wikisource,1.0
9,Aurora roja - Wikisource,1.0


 TF-IDF (Term Frequency - Inverse Document Frequency) es una herramienta matemática que sirve para descubrir qué palabras son verdaderamente importantes en un texto dentro de un conjunto de documentos, filtrando las que son comunes. 
 Se compone de dos partes:  
 TF (Frecuencia del término): Mide cuántas veces aparece una palabra dentro de ese mismo texto (si se repite mucho, probablemente sea central ahí).
 IDF (Frecuencia inversa de documento): Mide qué tan rara o común es esa palabra en todo el corpus. Si una palabra aparece en todos lados (como "el" o "de"), su valor baja; si aparece en pocos textos, es una palabra "rara" y su valor sube.  
 ¿Qué demuestran los parámetros por separado y juntos?
 1. Por separado: El efecto de smooth_idf=True (la primera tabla, en 1.0)
 Qué hace: En la teoría matemática tradicional, si una palabra aparece en el 100% de los documentos, su cálculo de IDF debería dar cero (lo que borraría por completo su existencia en el análisis).  
 Qué demuestra: Como scikit-learn aplica un sistema de suavizado interno por diseño, le suma un ajuste a la fórmula para que las palabras ultra comunes no mueran valiendo cero. 
 El resultado 1.0 en la tabla demuestra justamente eso: la librería establece un "piso mínimo" de importancia para que los términos transversales no desaparezcan. 

  Por separado: El efecto de la Norma L2 (Tu segunda tabla en 1.0)
    Qué hace: Agarra el vector numérico de cada documento y lo escala por su propia longitud para que el total de cada fila dé exactamente 1.0.
    Qué demuestra: Elimina el sesgo de longitud. Prueba que un libro de 500 páginas no va a tener más peso matemático que un cuento de 2 páginas solo por tener más palabras escritas; los pone a todos a competir en un mismo "escenario nivelado"

Combinados (El resultado final)
Juntos transforman un texto desordenado en un perfil matemático limpio y justo: smooth_idf evita que las palabras comunes se anulen por completo, y la norma L2 evita que los textos largos dominen a los cortos